json/json.isl

```diff
@@ -0,0 +1,51 @@
+type::{
+  name: json,
+  type: $any,
+  one_of: [
+    json_array,
+    json_boolean,
+    json_null,
+    json_number,
+    json_object,
+    json_string,
+  ],
+}
+
+type::{
+  name: json_array,
+  type: list,
+  annotations: closed::[],
+  element: json,
+}
+
+type::{
+  name: json_boolean,
+  type: bool,
+  annotations: closed::[],
+}
+
+type::{
+  name: json_null,
+  type: $null,
+  annotations: closed::[],
+}
+
+type::{
+  name: json_number,
+  one_of: [float, int],
```

Comment:
And decimals... they just have to be written without an exponent. Not sure how to model that here.

Reply:
That's what I though, at first.

However, I discovered that `ion-java` actually writes them with an exponent, and that the spec doesn't explicitly require a specific encoding for decimals when down-converting to json:

> Decimals are printed as JSON number with precision preserved 

So, it seems like decimals _could_ be down-converted without loss of information, but it depends on something that is left up to the particular Ion implementation that you're using.

Is my conclusion accurate?

Reply:
Let's take a step back. How is this schema intended to be used? Is it intended to validate that some text Ion data _already is_ valid JSON, or that some Ion data _could be_ re-encoded as valid JSON data?

If the latter, then I'd argue that we should just allow all decimals, as any decimal can be written without an exponent (using your example, `100d3` is `100000.`, which must be written as `100000` in JSON since it does not allow numbers to have a trailing `.`). When JSON down-conversion mode is on, all Ion writers should write decimals without exponents to make them valid JSON numbers. The ability to do this need not be implementation specific; it's inherent to the definition of a decimal.

Reply:
> Let's take a step back. How is this schema intended to be used? Is it intended to validate that some text Ion data already is valid JSON, or that some Ion data could be re-encoded as valid JSON data without data loss?

My intention was like the latter, but even stronger. The Ion _could_ be encoded as JSON and then up-converted back to Ion without any loss of information. (The motivation for this is that there are a lot of things [eg. frameworks, tools, databases] with JSON support, but not a lot with Ion support yet. It can be helpful to know/verify whether you can safely pass some Ion through as JSON and be able to get it back to Ion at the other end.)

I see now that I need to be more clear about that in the documentation.

> Any decimal can be written without an exponent (using your example, 100d3 is 100000., which must be written as 100000 in JSON since it does not allow numbers to have a trailing .). When JSON down-conversion mode is on, all Ion writers should write decimals without exponents to make them valid JSON numbers. The ability to do this need not be implementation specific; it's inherent to the definition of a decimal.

If `100d3` gets written out as `100000` and then later up-converted to Ion, it will be interpreted as an Integer (which I am considering a loss of information).

> Decimals are printed as JSON number with precision preserved ([source](https://amzn.github.io/ion-docs/guides/cookbook.html#down-converting-to-json))

Is precision used here in the mathematical sense (ie. number of sigfigs will be preserved) or is it saying that there will be no roundoff error caused by printing the number? (Or is it something else?) I presume it is the former because the [Ion Decimal spec](https://amzn.github.io/ion-docs/docs/decimal.html) specifically calls out that the representation preserves significant zeros.

If it is the former, then when `100d3` is written without an exponent, it will not preserve the number of significant zeros.

Reply:
> My intention was like the latter, but even stronger. The Ion could be encoded as JSON and then up-converted back to Ion without any loss of information.

Got it, thanks for the clarification. In that case, I agree that there are some decimals that cannot be roundtripped via JSON because they'd undergo a type change. There are some decimals that _can_ be roundtripped via JSON, but modeling that subset of decimals using Ion Schema would be much more complicated (assuming it's possible). As long as you're okay with the fact that this schema is conservative about what it allows, then it looks fine to me.

> If it is the former, then when 100d3 is written without an exponent, it will not preserve the number of significant zeros.

Yeah, you're right. You'd have to write it as `100e3` in order to accurately convey the precision, but that would be roundtripped as an Ion float, not decimal.